### PR TITLE
Expressions work better with strings.

### PR DESCRIPTION
**Pull request by gianm**
- ExpressionObjectSelector able to read from string columns, and able to
  return strings.
- ExpressionVirtualColumn able to offer string (and long for that matter)
  as its native type.
- ExpressionPostAggregator able to return strings.
- groupBy, topN: Allow post-aggregators to accept dimensions as inputs,
  making ExpressionPostAggregator more useful.
- topN: Use DimExtractionTopNAlgorithm for STRING columns that do not
  have dictionaries, allowing it to work with STRING-type expression
  virtual columns.
- Adjusts null handling to better match the rest of Druid: null and
  empty string treated the same; nulls implicitly treated as zeroes in
  numeric context.

### CODE REVIEW

File: processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java [hunk 1]
```diff
@@ -150,31 +148,28 @@ public String toString()
         .build();
   }
 
-  public static enum Ordering implements Comparator<Number>
+  public enum Ordering implements Comparator<Comparable>
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java]: Why enum?

File: processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java [hunk 2]
```diff
@@ -150,31 +148,28 @@ public String toString()
...
   {
     // ensures the following order: numeric > NaN > Infinite
     numericFirst {
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java]: Seems to be unused.

File: processing/src/main/java/io/druid/segment/virtual/ExpressionObjectSelector.java [hunk 3]
```diff
@@ -84,91 +88,61 @@ public static ExpressionObjectSelector from(ColumnSelectorFactory columnSelector
...
+        final Object val = selector.get();
+        if (val instanceof Number || val instanceof String) {
+          return val;
+        } else {
+          return null;
+        }
+      };
+    } else {
+      // No numbers or strings.
       return null;
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/virtual/ExpressionObjectSelector.java]: Should it be Supplier of null, not just null?

**Comment by fjy** [comment]
👍

File: processing/src/main/java/io/druid/segment/virtual/ExpressionVirtualColumn.java [hunk 4]
```diff
@@ -171,13 +178,14 @@ public boolean equals(final Object o)
     }
     final ExpressionVirtualColumn that = (ExpressionVirtualColumn) o;
     return Objects.equals(name, that.name) &&
-           Objects.equals(expression, that.expression);
+           Objects.equals(expression, that.expression) &&
+           outputType == that.outputType;
   }
 
   @Override
   public int hashCode()
```
Comment (leventov) [inline processing/src/main/java/io/druid/segment/virtual/ExpressionVirtualColumn.java]: Shouldn't `getCacheKey()` also be updated?

Comment (gianm) [inline processing/src/main/java/io/druid/segment/virtual/ExpressionVirtualColumn.java]: Yes, good catch. Updated it.

File: processing/src/main/java/io/druid/segment/virtual/ExpressionObjectSelector.java (hunk 3 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/virtual/ExpressionObjectSelector.java]: I guess it doesn't really matter, since a null supplier is treated equivalently to a supplier that always returns null.

File: processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java (hunk 2 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java]: It can be used via an `Ordering.valueOf(ordering)` in the constructor.

File: processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java (hunk 1 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java]: I'm not sure. It was already an enum and I didn't change it.

File: processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java [hunk 5]
```diff
@@ -36,25 +37,22 @@
...
-    {
-      if (o1 instanceof Long && o2 instanceof Long) {
-        return Long.compare(o1.longValue(), o2.longValue());
-      }
-      return Double.compare(o1.doubleValue(), o2.doubleValue());
-    }
-  };
+  private static final Comparator<Comparable> DEFAULT_COMPARATOR = Comparator.nullsFirst(
+      (Comparable o1, Comparable o2) -> {
+        if (o1 instanceof Long && o2 instanceof Long) {
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java]: I guess Druid isn't really consistent here. Sometimes nulls are treated as zeros (like when reading from nonexistent columns) and sometimes they are treated as less than any other number (like in most comparators).

My sense is that most comparators in Druid treat nulls as less than any other number (and actually return them to the user as `null` not as `0`) so I am inclined to keep this the way it is.

File: common/src/main/java/io/druid/math/expr/Expr.java [hunk 6]
```diff
@@ -57,7 +60,7 @@ default Object getLiteralValue()
 
   interface ObjectBinding
   {
-    Number get(String name);
+    Object get(String name);
```
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/Expr.java]: Annotated with `@Nullable`.

File: processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java (hunk 2 above)
Comment (leventov) [inline processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java]: It's too smart, suggested to change it to `NUMERIC_FIRST` constant with anonymous impl, and set it in constructor if the given ordering is `"numericFirst"`.

Also, this ordering is not null-proof, unlike the default comparator. It will throw NPE.

File: common/src/main/java/io/druid/math/expr/Parser.java [hunk 7]
```diff
@@ -148,30 +148,14 @@ public void visit(Expr expr)
...
-      public Number get(String name)
-      {
-        Number number = (Number)bindings.get(name);
-        if (number == null && !bindings.containsKey(name)) {
-          throw new RuntimeException("No binding found for " + name);
-        }
-        return number;
-      }
-    };
+    return bindings::get;
```
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/Parser.java]: Because rather than treating expressions referencing missing columns as errors, I'd like to treat them as if they were reading nulls. Just doing `bindings::get` achieves that since `bindings.get(nonexistentColumnName) == null`

File: common/src/main/java/io/druid/math/expr/ExprEval.java [hunk 8]
```diff
@@ -242,25 +242,37 @@ public final ExprType type()
...
     @Override
     public final int asInt()
     {
-      return Integer.parseInt(value);
+      if (value == null) {
+        return 0;
+      }
+
+      final Integer theInt = Ints.tryParse(value);
+      return theInt == null ? 0 : theInt;
```
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/ExprEval.java]: I thought it would be more useful to treat unparseable strings as zeroes rather than to throw errors. I don't want one bad string in a column to make the entire query fail. It's similar to the behavior of Druid's "numeric" comparator on string columns, where unparseable numbers as less than all other numbers, not as errors.

File: common/src/main/java/io/druid/math/expr/Parser.java (hunk 7 above)
Comment (leventov) [inline common/src/main/java/io/druid/math/expr/Parser.java]: There is a tradeoff between "ease of use" and the problems from consuming client errors. Maybe it should be a configurable?

Comment (gianm) [inline common/src/main/java/io/druid/math/expr/Parser.java]: I don't think this needs to be configurable for expressions, since the behavior of treating nonexistent columns as having some "default" value is very ingrained in Druid -- not just in the expressions but throughout the system in general.

I guess you could argue that you want it to be configurable throughout Druid, but that is out of scope for this PR, imo.

File: common/src/main/java/io/druid/math/expr/Expr.java [hunk 9]
```diff
@@ -527,7 +528,7 @@ protected final long evalLong(long left, long right)
   @Override
   protected final double evalDouble(double left, double right)
   {
-    return Evals.asDouble(left < right);
+    return Evals.asDouble(Double.compare(left, right) < 0);
```
Comment (leventov) [inline common/src/main/java/io/druid/math/expr/Expr.java]: Could you please comment on this and similar places, not to amuse later readers why such "long" form is chosen over "simple" `left < right`.

File: common/src/main/java/io/druid/math/expr/ExprEval.java [hunk 10]
```diff
@@ -242,25 +242,37 @@ public final ExprType type()
     @Override
     public final boolean isNull()
     {
-      return Strings.isNullOrEmpty(value);
+      return value == null;
```
Comment (jihoonson) [inline common/src/main/java/io/druid/math/expr/ExprEval.java]: This method is same with the overridden method and thus can be removed.

File: common/src/main/java/io/druid/math/expr/Expr.java (hunk 9 above)
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/Expr.java]: Yes, I'll add comments.

File: processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java (hunk 2 above)
Comment (gianm) [inline processing/src/main/java/io/druid/query/aggregation/post/ExpressionPostAggregator.java]: The ordering is null-proof. If either lhs or rhs is null then the final else branch will be taken, which uses `Comparators.naturalNullsFirst()`.

I don't really have strong feelings about how the constructor should be implemented, but I don't want to change it in this PR, since it's not relevant to the main purpose. NB: If you (or anyone else) ends up reading this and doing another patch to change it, a very similar construction is used in ArithmeticPostAggregator.

I added a comment though explaining how the constructor works.

File: common/src/main/java/io/druid/math/expr/ExprEval.java (hunk 10 above)
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/ExprEval.java]: Thanks, removed.

File: processing/src/main/java/io/druid/query/Queries.java [hunk 11]
```diff
@@ -45,6 +45,7 @@
   }
 
   public static List<PostAggregator> prepareAggregations(
+      List<String> otherOutputNames,
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/Queries.java]: I'll add javadocs.

File: common/src/main/java/io/druid/math/expr/ExprEval.java [hunk 12]
```diff
@@ -150,7 +150,7 @@ public final double asDouble()
   {
     private DoubleExprEval(Number value)
     {
-      super(value);
+      super(Preconditions.checkNotNull(value, "value"));
```
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/ExprEval.java]: The value can't actually be null. For reasons of trying to make the behavior more consistent with Druid null handling in general, `null` value is always going to be a `StringExprEval`. (See that the constructors for DoubleExprEval and LongExprEval have a non-null check for value) I'll remove the null checks in toExpr, since they are pointless, given value cannot be null.

File: common/src/main/java/io/druid/math/expr/ExprEval.java [hunk 13]
```diff
@@ -190,7 +190,7 @@ public Expr toExpr()
   {
     private LongExprEval(Number value)
     {
-      super(value);
+      super(Preconditions.checkNotNull(value, "value"));
```
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/ExprEval.java]: Same response as https://github.com/druid-io/druid/pull/4394#discussion_r121711016.

**Comment by gianm** [comment]
@leventov, @jihoonson, thanks for reviewing. I pushed a new update.

**Comment by b-slim** [comment]
@gianm i think there is some tests that needs some fixes 
```java
GroupByQueryRunnerTest.testGroupByWithOutputNameCollisions[config=v2SmallDictionary, runner=noRollupRtIndex] 

Expected: (an instance of java.lang.IllegalArgumentException and exception with message a string containing "Duplicate output name[alias]")

     but: exception with message a string containing "Duplicate output name[alias]" message was "[alias] already defined"

Stacktrace was: java.lang.IllegalArgumentException: [alias] already defined
```

**Comment by gianm** [comment]
@b-slim ah, yeah, one of the error messages changed. I just pushed a fix for that, thanks.

**Comment by fjy** [comment]
@b-slim any more comments?

File: common/src/main/java/io/druid/math/expr/ExprEval.java (hunk 12 above)
Comment (jihoonson) [inline common/src/main/java/io/druid/math/expr/ExprEval.java]: I see. Then, would you remove ```@Nullable``` annotations at [DoubleExpr.getLiteralValue()](https://github.com/druid-io/druid/blob/master/common/src/main/java/io/druid/math/expr/Expr.java#L155) and [LongExpr.getLiteralValue()](https://github.com/druid-io/druid/blob/master/common/src/main/java/io/druid/math/expr/Expr.java#L95)?

Comment (gianm) [inline common/src/main/java/io/druid/math/expr/ExprEval.java]: Yes, ok. I removed those and added non-null checks to the constructors.

**Comment by gianm** [comment]
@fjy are you ok with the design here?

**Comment by fjy** [comment]
yeah +1

**Comment by gianm** [comment]
Ok, I think it's mergeable then, since we have 3 +1s and one of the reviewers looked at the code. Thanks everyone.

File: processing/src/main/java/io/druid/query/Queries.java [hunk 14]
```diff
@@ -44,34 +47,57 @@
...
+   * @param aggFactories     aggregator factories for this query
+   * @param postAggs         post-aggregators for this query
+   *
+   * @return decorated post-aggregators
+   *
+   * @throws NullPointerException     if otherOutputNames or aggFactories is null
+   * @throws IllegalArgumentException if there are any output name collisions or missing post-aggregator inputs
+   */
   public static List<PostAggregator> prepareAggregations(
+      List<String> otherOutputNames,
```
Comment (leventov) [inline processing/src/main/java/io/druid/query/Queries.java]: Is `Queries` considered public API? Because this is a breaking change. E. g. there is a usage of this method in our extensions.

Comment (gianm) [inline processing/src/main/java/io/druid/query/Queries.java]: I'm not sure, good question. I guess we might as well consider it public, since it's likely that people will have copied it from builtin query impls.

Comment (gianm) [inline processing/src/main/java/io/druid/query/Queries.java]: #4432
